### ISSUE TITLE
Team Management - Assigning teams to plan-jurisdictions

### ISSUE THREAD

**Issue by p-netm** [onaio/reveal-frontend#411]
__Overview__

**Comment by cKellyDesign** [comment]
@p-netm do you have sufficient scoping for this?

**Comment by moshthepitt** [comment]
@cKellyDesign This work was just the UI.  When the API is ready, we'll probably have a different issue to connect that to the UI.

**Comment by p-netm** [comment]
Proposed workflow:
Assuming that for a plans page we already have the plan_id for that plan and all jurisdiction_ids mapped to that plan.

1. create a multiJurisdictionSelect component .
2. Place that inside the teamform. - will use that to assign at least one jurisdiction to a team.
3. From the plan page clicking the assign team button; should show the user a dropdown with list of teams from which they can pick from
    - the teams the user can pick from will only be those that have at least one jurisdiction that is a sibling or a children of one the plan's jurisdictions?!!!

@moshthepitt @cKellyDesign

**Comment by moshthepitt** [comment]
> create a multiJurisdictionSelect component .

How will this one work?

**Comment by p-netm** [comment]
> > create a multiJurisdictionSelect component .
> 
> How will this one work?

not quite sure, but i was thinking of having something like the jurisdictionselect component but with the ability to pick more than a single jurisdiction. at a time, using the `isMulti` select prop

**Comment by moshthepitt** [comment]
Since this is a drill down select, in case one needs to be able to select at any level of the hierarchy, the UX/flow for this becomes challenging.  We definitely need input from @rowo in that case.

**Comment by p-netm** [comment]
ok, will leave this hanging and effectively blocked untill then.

**Comment by rowo** [comment]
@moshthepitt @cKellyDesign @p-netm Does a "checkable tree" work for what you are talking about here? See http://jonmiles.github.io/bootstrap-treeview/ (search "checkable tree")

<img width="752" alt="Screen Shot 2019-09-23 at 12 27 48 AM" src="https://user-images.githubusercontent.com/1584163/65401909-011e2b80-dd99-11e9-8a7a-a6e8a24e6495.png">

As a dialog, this checkbox set up would need buttons for "Save" and "Cancel". The dialog can have a title or not. I guess if it does it would say "Assign team" or "Assign team to <operational area>". 

Main thing is this should be done in most minimal way possible. So if there's another way you can think of that is easier, let me know. For example, would it be faster to do a select-one pulldowns with a tree and then do a "add" button to add more location? Or perhaps do a set of pulldowns from top to bottom hierarchy where it saves whatever level you stop selecting on.

**Comment by mberg** [comment]
We aren't assigning teams to jurisdictions in the team UI.  We're
assigning it in the Plan assignment.

Idea is you click on a jurisdiction and select the teams assigned to that
jurisdiction.  This is the same as the mockups except the that you can
assign multiple teams not just one.

**Comment by moshthepitt** [comment]
> Idea is you click on a jurisdiction and select the teams assigned to that jurisdiction. 

@mberg we've been stuck on this for a while hence the above discussion.  Regarding **teams assigned to that jurisdiction** - we need to do this in the web UI somewhere right?  Or else we'll still have to use OpenMRS and we will not support multi-jurisdiction assignment. Right?

**Comment by mberg** [comment]
Need to confirm with Sam but the data model has changed.

We just need to assign teams to plan -> locations.  We don't need to assign teams to locations anymore.

**Comment by moshthepitt** [comment]
@mberg @rowo thanks for the input.  Things are now clearer.  We'll ping you when we have something to showcase.

**Comment by cKellyDesign** [comment]
I think @p-netm's approach can be applied to the popup where we are selecting Teams via the [AsynSelect component](https://github.com/JedWatson/react-select) we're already using throughout the site. With a bit of styling it can probably match the mockups pretty easily.

**Comment by cKellyDesign** [comment]
@moshthepitt @p-netm @pld - we probably don't want to trigger anything **actual** assignment until the user is good and ready to submit the assignments (to avoid triggering unnecessary flows / task generating / task updating / etc - Correct me if this isn't a concern)

@rowo @OluwatobiMayowa ☝️ Given this assumption, should we have `Save Assignments as Draft` and `Finalize and Assign` buttons which (trigger actual API calls and) do the business of assigning teams to plan-jurisdictions?

**Comment by cKellyDesign** [comment]
@githengi @Rkareko @pld - How do we save these `organization-to-planJurisdiction` assignments?

- [ ] What endpoint do we use?
- [ ] What data/schema do we include in the payload?
- [ ] Once created, can we just `PUT` instead of `POST` to update `organization-to-planJurisdiction` assignments?

**Comment by cKellyDesign** [comment]
Another question for @rowo and @OluwatobiMayowa: Where does this workflow start and end? In the mockups the flow [starts immediately after finalizing a draft plan](https://marvelapp.com/b2iaahb/screen/53843108) and [doesn't really have an end](https://marvelapp.com/b2iaahb/screen/53843113).

- [ ] How can a client get to this flow for an existing Finalized plan? (where do we put the button?)
- [ ] After finalizing and saving team assignments, what should happen? (where do we navigate?)

**Comment by pld** [comment]
In this case what's the difference btwn Draft and Finalize? Can we combine those states?

**Comment by cKellyDesign** [comment]
> what's the difference btwn Draft and Finalize

@pld Great question, I don't know the various backend dependencies or triggers around switching a Plan status from `draft` to `status`. Oh, one thing would be the fact that we are assigning teams to `plan-definition` entities which, definitionally, don't exist until a plan exists with jurisdictions. 

@p-netm @githengi @moshthepitt - are we able to get plan-jurisdiction entities directly from OpenSRP or is this something we are currently only getting from Superset slices? If the later, then another issue will be the time delay between saving things and seeing them populated in Superset.

From a UI perspective, we do currently have space to add a column which can include a list of assigned teams and a button to add and update, which triggers the popup. This would mean the user can select jurisdictions for the plan and assign teams to that plan-jurisdiciton at the same time (@rowo @mberg - thoughts on this?).

**Comment by githengi** [comment]
> are we able to get plan-jurisdiction entities directly from OpenSRP 
This is currently supported @cKellyDesign

**Comment by cKellyDesign** [comment]
> This is currently supported @cKellyDesign

@githengi great thanks! when/how are plan-jurisdiction entities created in opensrp?

**Comment by moshthepitt** [comment]
@cKellyDesign sorry I don't have anything else to add apart from "lets avoid using Superset at all costs for this".

plan-jurisdictions can be inferred from the jurisdictions field in plab objects

**Comment by moshthepitt** [comment]
> @moshthepitt @p-netm @pld - we probably don't want to trigger anything actual assignment until the user is good and ready to submit the assignments (to avoid triggering unnecessary flows / task generating / task updating / etc - Correct me if this isn't a concern)

I do not think assigning teams does trigger any flows/events/other things/etc @Wambere correct me here if I am wrong.

**Comment by cKellyDesign** [comment]
> plan-jurisdictions can be inferred from the jurisdictions field in plab objects

Cool yeah we can infer to render the page correctly, ultimately we will need to know whether to `POST` or `PUT` the api call, so we'll need to know by that point if the plan-jurisdiction exists or not.

> I do not think assigning teams does trigger any flows/events/other things/etc

Excellent, then we can assign in Draft or Finalized sate, seems like doing it in Draft makes more sense and we can assign teams to plan-jurisdictions in chunks when the user clicks `Save as Draft` and `Finalize Plan` (to avoid excess API calls during the process of actually choosing teams)

**Comment by pld** [comment]
> @githengi @Rkareko @pld - How do we save these `organization-to-planJurisdiction` assignments?
> 
>     * [ ]  What endpoint do we use?

```
POST
https://reveal-[ENV].smartregister.org/opensrp/rest/organization/assignLocationsAndPlans
```

>     * [ ]  What data/schema do we include in the payload?

This expects an array of objects, e.g. for a single assignment:

```json
[
  {
    "jurisdiction": "f8f89aff-056a-4322-9fe0-db5501f51552",
    "organization": "66c11ebf-623c-401e-87be-e4f38378c18c",
    "plan": "9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d",
    "fromDate": "2019-09-12T00:00:00.000Z",
    "toDate": "2019-12-01T00:00:00.000Z"
  }
]
``` 

I've heard that it should be able to take multiple objects in that array, but the last time I tried that (Wednesday late night) it only succeeded when the array had a single object, as above.

>     * [ ]  Once created, can we just `PUT` instead of `POST` to update `organization-to-planJurisdiction` assignments?

`POST` works, I have not tried `PUT` if this becomes blocking, you should test and see, would also be good to know if `DELETE` works on this endpoint, I haven't tried anything other than `POST`

@githengi please let us know what the API is for this.

**Comment by githengi** [comment]
@cKellyDesign please check this documentation as a starting point.

You use a POST to https://reveal-[ENV].smartregister.org/opensrp/rest/organization/assignLocationsAndPlans with a list of assignment objects as @pld has described above.

Currently PUT or DELETE are not supported.

However POST allows for updates if the organization, plan and jurisdiction combination exists and has not expired, otherwise it creates the assignment as new.

To delete an assignment you will send the "toDate" with the date when its not longer valid.

This was developed with the concept or roving teams in mind such that once can define several assignments that will expire and new ones can kick in.

**Comment by cKellyDesign** [comment]
> This was developed with the concept or roving teams in mind such that once can define several assignments that will expire and new ones can kick in.

Ok then when creating these assignments we'll set the `toDate` to match the plan's end-date.

@githengi @pld - is it possible to get a list of organizations by `plan` and `jurisdiction` parameters? Currently [it doesn't look like it](https://github.com/OpenSRP/opensrp-server-web/blob/master/src/main/java/org/opensrp/web/rest/OrganizationResource.java#L50-L58), or are we able to filter by params like other endpoints?

**Comment by pld** [comment]
Not that I'm aware of, I see a properties filter parameter in the `LocationsResource`, not sure if that's helpful or if it allows the relevant filtering.

If not let's parse the all orgs returned JSON for now and work with Sam to add an appropriate ticket on OpenSRP server for more performant querying.

**Comment by cKellyDesign** [comment]
Issue created at OpenSRP/opensrp-server-web#148 to get Assignments by planId so that we can render existing Assignments when loading draft plans.

**Comment by cKellyDesign** [comment]
Removing `blocked` label for now, currently doing work on the UI and the local Store and will worry about `GET`ing assignments from OpenSRP a little later.

**Comment by cKellyDesign** [comment]
@githengi @pld when `POST`ing to the endpoint with the following payload:
```
[
  {
    "jurisdiction": "f8f89aff-056a-4322-9fe0-db5501f51552",
    "organization": "66c11ebf-623c-401e-87be-e4f38378c18c",
    "plan": "9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d",
    "fromDate": "2019-09-12T00:00:00.000Z",
    "toDate": "2019-12-01T00:00:00.000Z"
  }
]
```

Should `jurisdiction` always be an Operation/Focus/Spray Area or should we also be generating assignment objects for Administrative Jurisdictions?

**Comment by pld** [comment]
Good question, I unfortunately don't know the answer to that, @mberg may also have knowledge on this

**Comment by githengi** [comment]
@cKellyDesign for now it has to be Operation/Focus/Spray Area, In future was can support assignment at higher levels

**Comment by cKellyDesign** [comment]
@mberg suggested keeping it simple for now and simply requiring the user to drill down to the Spray Areas and assigning one at a time.

**Comment by cKellyDesign** [comment]
@pld @githengi I'm successfully `GET`ing from the `assignedLocationsAndPlans` endpoint, but it seems I can no longer `POST` to this endpoint. ([the resource](https://github.com/OpenSRP/opensrp-server-web/blob/master/src/main/java/org/opensrp/web/rest/OrganizationResource.java#L152))

**Comment by githengi** [comment]
The Post endpoint should be assignLocationsAndPlans

**Comment by cKellyDesign** [comment]
@githengi @pld @Rkareko - it looks like when `GET`ing the assignments they are not including the `organization` identifiers.

POST'd payload:
```
[
      {
              jurisdiction: 'b777f3a1-de51-4431-a9a4-2b1bfd4ba852',
              organization: 'fcc19470-d599-11e9-bb65-2a2ae2dbcce4',
              plan: '7c941be5-ff00-5ce0-a21a-61f761a1a8df',
              fromDate: '2019-10-06T00:00:00-07:00',
              toDate: '2019-11-30T00:00:00-08:00',
      },
      {
              jurisdiction: 'b777f3a1-de51-4431-a9a4-2b1bfd4ba852',
              organization: '124ac19a-8eaf-479a-b937-1cb4779d7314',
              plan: '7c941be5-ff00-5ce0-a21a-61f761a1a8df',
              fromDate: '2019-10-06T00:00:00-07:00',
              toDate: '2019-11-30T00:00:00-08:00',
      },
      {
              jurisdiction: '6fe7bf41-132d-4756-989a-42b95732a522',
              organization: '2dd255b7-3f40-5ab0-95da-f91ff4fd03b8',
              plan: '7c941be5-ff00-5ce0-a21a-61f761a1a8df',
              fromDate: '2019-10-06T00:00:00-07:00',
              toDate: '2019-11-30T00:00:00-08:00',
      },
]
```
GET Results in:
```
[
      {
            "jurisdictionId": "b777f3a1-de51-4431-a9a4-2b1bfd4ba852",
            "planId": "7c941be5-ff00-5ce0-a21a-61f761a1a8df",
            "fromDate": 1570312800000,
            "toDate": 1575064800000
      },
      {
            "jurisdictionId": "b777f3a1-de51-4431-a9a4-2b1bfd4ba852",
            "planId": "7c941be5-ff00-5ce0-a21a-61f761a1a8df",
            "fromDate": 1570312800000,
            "toDate": 1575064800000
      },
      {
            "jurisdictionId": "6fe7bf41-132d-4756-989a-42b95732a522",
            "planId": "7c941be5-ff00-5ce0-a21a-61f761a1a8df",
            "fromDate": 1570312800000,
            "toDate": 1575064800000
      }
]
```

**Comment by pld** [comment]
Yeah that's correct you query with the org ID

**Comment by cKellyDesign** [comment]
@pld This is querying with the plan Id.

**Comment by pld** [comment]
Oh I see, I'd guess the template for the return value was never updated

**Comment by moshthepitt** [comment]
fixed by #503
